### PR TITLE
Weave by timestamp

### PR DESCRIPTION
**Pull request by mnshaw**
Add option to weave logs together by timestamp.

@timstclair @rmmh

### CODE REVIEW

File: gubernator/timestamp_parser.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import regex
+
+def sort_by_timestamp(lines):
+    # Given an array of log lines, returns an array of the lines sorted by timestamp
+    timestamp_dict = {}
```
Comment (rmmh) [inline gubernator/timestamp_parser.py]: This should be something closer to `return sorted(lines, key=regex.extract_timestamp)` -- don't build the intermediate structure like this. And if it's that simple, it doesn't need its own file.

File: gubernator/regex.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ def containerID(line):
     return re.search(r'ContainerID:([0-9A-Fa-f]*)', line)
 
 def timestamp(line):
-    return re.search(r'(\d\d:\d\d:\d\d\.\d+)', line)
-
-def clean_timestamp(line):
-    return re.sub(r'(:|\.)', "", line)
+    timestamp_re = re.search(r'(\d\d:\d\d:\d\d\.\d+)', line)
+    if timestamp_re and timestamp_re.group(0):
+        return re.sub(r'(:|\.)', "", timestamp_re.group(0))
```
Comment (rmmh) [inline gubernator/regex.py]: why is this necessary? It will sort properly with `:` and `.`

File: gubernator/regex.py [hunk 3]
```diff
@@ -41,7 +41,7 @@ def containerID(line):
     return re.search(r'ContainerID:([0-9A-Fa-f]*)', line)
 
 def timestamp(line):
-    return re.search(r'(\d\d:\d\d:\d\d\.\d+)', line)
-
-def clean_timestamp(line):
-    return re.sub(r'(:|\.)', "", line)
+    timestamp_re = re.search(r'(\d\d:\d\d:\d\d\.\d+)', line)
```
Comment (rmmh) [inline gubernator/regex.py]: Can you include days in the regex as well to handle wraparound at midnight?

File: gubernator/log_parser.py [hunk 4]
```diff
@@ -75,15 +75,23 @@ def log_html(lines, matched_lines, hilight_words, skip_fmt):
...
+def digest(data, objref_dict=None, filters=None, error_re=regex.error_re,
+    skip_fmt=lambda l: '... skipping %d lines ...' % l, timestamps=None):
+    # pylint: disable=too-many-arguments
     """
     Given a build log, return a chunk of HTML code suitable for
     inclusion in a <pre> tag, with "interesting" errors hilighted.
 
     This is similar to the output of `grep -C4` with an appropriate regex.
     """
     lines = unicode(jinja2.escape(data)).split('\n')
```
Comment (timstclair) [inline gubernator/log_parser.py]: Remove this line

File: gubernator/templates/filtered_log.html [hunk 5]
```diff
@@ -13,6 +13,7 @@
 		<input type="hidden" name="junit" value="{{junit}}">
 		<hr>
 		<label><input type="checkbox" name="wrap"{% if wrap %} checked{% endif %}> Line wrapping on</label>
+		<label><input type="checkbox" name="weave"{% if weave %} checked{% endif %}> Weaving by timestamp on</label>
```
Comment (timstclair) [inline gubernator/templates/filtered_log.html]: nit: I'd drop the "on" from the labels

**Comment by timstclair** [comment]
I didn't totally follow the code, but are you including the build log as well? The build log is often generated outside the cluster and uses a different time zone, so it will be offset. I think it is safe to assume all the cluster logs use the same timezone.

Also, I don't think you handle multi-line log messages? To handle multiline messages, I recommend joining the lines without timestamps into the previous line with a timestamp. That might be weird with the build log though (maybe it's best to just keep the build log separate?)

File: gubernator/log_parser.py (hunk 4 above)
Comment (rmmh) [inline gubernator/log_parser.py]: Don't remove that line, it's essential.

File: gubernator/log_parser.py [hunk 6]
```diff
@@ -75,15 +75,23 @@ def log_html(lines, matched_lines, hilight_words, skip_fmt):
...
+    if timestamps:
```
Comment (rmmh) [inline gubernator/log_parser.py]: You should do the combining/sorting outside of digest, and only ever pass a string to digest. Having the type of the arguments depend on a bool is confusing.

**Comment by mnshaw** [comment]
@timstclair No, the build-log.txt file is not one of the files that can be filtered right now. I think right now it is only used by Gubernator when it can't find the failures of the build from the junit files.
Combining the multiline messages sounds good.

File: gubernator/regex.py [hunk 7]
```diff
@@ -41,7 +41,7 @@ def containerID(line):
...
 
 def timestamp(line):
-    timestamp_re = re.search(r'(\d\d-?\d\d[T\s]\d\d:\d\d:\d\d\.\d+)', line)
-    if timestamp_re and timestamp_re.group(0):
-        return re.sub(r'(-|T|\s)', "", timestamp_re.group(0))
-    return ""
+    return re.search(r'(\d\d-?\d\d[T\s]\d\d:\d\d:\d\d\.\d+)', line)
+
+def sub_timestamp(line):
+    return re.sub(r'(-|T|\s)', "", timestamp(line).group(0))
```
Comment (rmmh) [inline gubernator/regex.py]: Is this necessary for lexicographic ordering? Is there a mix of timestamps of slightly different formats that this normalizes?

Comment (mnshaw) [inline gubernator/regex.py]: I might be missing some, but these are examples of two types of timestamps I've seen, which is why I had that line:
I0629 17:33:09.813041
2016-07-22T19:01:11.150204523Z

Comment (rmmh) [inline gubernator/regex.py]: Neat. Can you add a regex test explaining that and demonstrating how the two different formats can normalize to the same thing?

File: gubernator/view_logs.py [hunk 8]
```diff
@@ -167,6 +168,48 @@ def get_logs(build_dir, log_files, pod_name, filters, objref_dict):
...
+    combined_lines = []
+    first_combined = ""
+    pod_re = regex.wordRE(pod)
+
+    for log_file in log_files:
+        log = gcs_async.read(log_file).get_result()
+        logs.append(log.decode('utf8', 'replace'))
+    for log in logs:
+        lines.extend(log.split('\n'))
+    for line in lines:
```
Comment (rmmh) [inline gubernator/view_logs.py]: Please comment on the purpose of this for loop.

File: gubernator/view_logs.py [hunk 9]
```diff
@@ -167,6 +168,48 @@ def get_logs(build_dir, log_files, pod_name, filters, objref_dict):
...
+    lines = []
+    combined_lines = []
+    first_combined = ""
+    pod_re = regex.wordRE(pod)
+
+    for log_file in log_files:
+        log = gcs_async.read(log_file).get_result()
+        logs.append(log.decode('utf8', 'replace'))
+    for log in logs:
+        lines.extend(log.split('\n'))
```
Comment (rmmh) [inline gubernator/view_logs.py]: Possible TODO: is it useful to indicate which log file a line came from?

Comment (mnshaw) [inline gubernator/view_logs.py]: I guess that information could be used to somehow differentiate the lines when they're woven together? For example the text could be in different colors for different logs... I don't know how useful that would be.

**Comment by rmmh** [comment]
Please make a test to demonstrate the end-to-end functionality-- none of the tests are executing get_woven_logs or parse_by_timestamp

You can see this by doing:

    ./test.sh --with-coverage --cover-erase --cover-inclusive --cover-package gubernator --cover-html
    google-chrome cover/index.html

File: gubernator/templates/filtered_log.html [hunk 10]
```diff
@@ -12,7 +12,8 @@
 		</h3>
 		<input type="hidden" name="junit" value="{{junit}}">
 		<hr>
-		<label><input type="checkbox" name="wrap"{% if wrap %} checked{% endif %}> Line wrapping on</label>
+		<label><input type="checkbox" name="wrap"{% if wrap %} checked{% endif %}> Line wrapping </label>
+		<label><input type="checkbox" name="weave"{% if weave %} checked{% endif %}> Weaving by timestamp </label>
```
Comment (timstclair) [inline gubernator/templates/filtered_log.html]: nit: UI generally uses the imperative tense (e.g. `Wrap lines` or `Weave by timestamp`)

File: gubernator/templates/filtered_log.html [hunk 11]
```diff
@@ -27,10 +28,16 @@
...
-		<pre {% if not wrap %} style="white-space:pre"{% endif %}>
-			{{logs[file] | safe}}
+	% if weave
+		<pre {% if not wrap %} style="white-space:pre"{% endif %}> 
+			{{woven_logs | safe}}
 		</pre>
-	% endfor
+	% else
+		% for file in log_files|sort
+			<h3>{{file}}</a></h3>
```
Comment (timstclair) [inline gubernator/templates/filtered_log.html]: I think that `</a>` doesn't belong?

**Comment by rmmh** [comment]
LGTM. Feel free to squash and merge.
